### ISSUE TITLE
EVENTS TSTOP in CTA DC-1

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#1866]
I think in CTA DC-1 the `TSTOP` in the `EVENTS` header isn't filled 100% correctly.

For this AGN run it should be 15 min = 900 seconds after TSTART, but it's 4 seconds earlier:
```
>>> from astropy.io import fits
>>> hdu_list = fits.open("/Users/deil/work/data/cta/1dc/1dc/data/baseline/agn/agn_baseline_510000.fits")
>>> hdu_list["EVENTS"].header["TSTART"]
662774400.0
>>> hdu_list["EVENTS"].header["TSTOP"]
662775296.0
>>> hdu_list["EVENTS"].header["TSTOP"] - hdu_list["EVENTS"].header["TSTART"]
896.0
```
These 4 seconds are half the dead-time correction for this run, so presumably the issue is related to `DEADC`:
```
>>> hdu_list["EVENTS"].header["DEADC"]
0.98000001907
```

This does affect Gammapy, because at the moment we access `TSTOP` from the `EVENTS` header to get the run stop time. This results in events up to 4 seconds after the run stop in that case:

```
>>> from gammapy.data import DataStore
>>> ds = DataStore.from_dir('1dc/1dc/index/agn')
>>> obs = ds.obs(510000)
>>> obs.events.observation_time_end
<Time object: scale='tt' format='mjd' value=59215.52074074074>
>>> obs.events.time.max()
<Time object: scale='tt' format='mjd' value=59215.520832720315>
>>> (obs.events.time.max() - obs.events.observation_time_end).sec
3.9984025955200844
```

Checking all CTA DC-1 AGN runs I find deviations up to plus or minus 32 seconds:
![cta_dc1_data_time_diff](https://user-images.githubusercontent.com/852409/46586337-735d0980-ca7d-11e8-8283-9204b0afe1fc.png)

Note that last week we changed the IRF background computation to be per ONTIME = TSTOP - TSTART, following the spec (see #1861 , cc @registerrier ). Also we are still investigating some differences in background between Gammapy and the simulation from ctobssim, which might be partly due to issues with the run time definition (see https://github.com/gammapy/gammapy/issues/1842#issuecomment-427679687 and CC @facero and @adonath ).

I didn't check all runs yet, but in the AGN run 510000  mentioned above, GTI is filled correctly:
```
>>> hdu_list["GTI"].data["STOP"] - hdu_list["GTI"].data["START"] 
array([900.])
```
So if we change exposure and time computations in Gammapy to be GTI based instead of EVENTS header key based (cc @dcfidalgo), results for DC-1 will change a bit; probably at the 1% level, or even less in many cases, I don't think this is a major issue, just a small one.

I've asked Jürgen via email to check if this is an issue in ctobssim, filling TSTOP incorrectly, and for background IRF whether they apply DEADC there or not for the rate computations.

If someone has time to write a notebook that checks EVENT TSTOP against GTI TSTOP for all data (and maybe also check TSTART for consistency, or check LIVETIME against ONTIME and DEADC), please go ahead and link to it from this issue. Otherwise I can do this, assigning to me for now.

**Comment by cdeil** [comment]
I did a check: https://nbviewer.jupyter.org/gist/cdeil/a3c798ebefe4cd2bb79f74c58f1c85bb

Basically in DC-1 the TSTART and TSTOP in the EVENTS header are off a bit, and the GTI contains the correct values. I've sent another email to Jürgen about this.

@dcfidalgo @registerrier and maybe @cboisson - How about we change gammapy.data to always use GTI? (Instead of a mix of sometimes using time-related infos from the EVENTS header.) It has the effect to make GTI required, but maybe that's OK and the best solution here?

I'm slightly +1 to this, irrespective of the issue with the DC-1 files. For DC-1, we can just take this as motivation to implement more flexible background modeling in Gammapy, like a norm that is fitted to absorb this effect at the ~ 1-2% level.

**Comment by cdeil** [comment]
On the technical side, I think we need to change the Gammapy data access code to load the EVENTS header only via `astropy.io.fits.open('events.fits')['EVENTS'].header`, i.e. without loading the EVENTS table data, and to load the GTI table, for efficiency reasons. E.g. to compute exposure, it's acceptable to load the EVENTS header or the small GTI table, but we should avoid loading the EVENTS table data.

For this part we could introduce an `ObservationInfo` object to represent and load the infos like pointing and times. In HESS we have such a `RunInfo` object. Or we could just change the `Observation` class to load that part either on init or first access. That part of the observation info should probably be cached on the observation object.

To me, this is still quite fuzzy how exactly this should work. @dcfidalgo - Can you come with a proposal or a PR in that direction? Or should we first discuss a bit, e.g. at one of the next Gammapy calls, and try to develop a solution together?

**Comment by dcfidalgo** [comment]
Hi,

> How about we change gammapy.data to always use GTI? (Instead of a mix of sometimes using time-related infos from the EVENTS header.) It has the effect to make GTI required, but maybe that's OK and the best solution here?

I am also +1 for always using the GTI table when computing the ontime, i think it is definitely ok to require a GTI table.

About caching some observation info, i guess it is definitely a good idea, we already do it at the moment. Maybe for the new `Observation` class it will become slightly more complicated since we want to apply filters on the fly. That is, for example, if the filter for the time selection changes, the cached values have to be updated as well.

Since all this new `Observation` class thing is still not fully worked out, i suggest to first implement a straight forward solution (not caching anything) and then start to think how to optimize it and to make it more efficient. I hope i manage to distribute a first version of the Data handling PIG still today.

**Comment by cdeil** [comment]
Resolved by @dcfidalgo in #1908 .
